### PR TITLE
Support for unwrapping key via an HSM when decrypting the SAML assertion.

### PR DESCRIPTION
**Pull request by RobertButtigieg**
This pull request addresses [#257](https://github.com/onelogin/java-saml/issues/257)

Currently, the library has no support for HSM integration, in particular when it comes to unwrap the key as part of the SAML decryption process.

Hence, this solution will support HSM key unwrapping by setting an HSM object as part of the SAML settings. My requirement was to integrate this library specifically to the Azure Key Vault. However, I have implemented an HSM abstract class in order to allow the possibility of integrating it with other HSMs as well. I have also imported 2 new libraries (for the Azure Key Vault) as optional dependencies since whoever is going to use this library might not require this HSM support. 

I have also upgraded the dependency check plugin to the current latest version (v6.0.3) since it includes fixes of the false positives which were flagged when those 2 new libraries were imported.

### CODE REVIEW

**Comment by ArjanSchouten** [comment]
I don't want and need Azure dependencies in my codebase.
So I'm not happy when this is merged. If this is relevant to some people, isn't it better to implement this as a separate module?

**Comment by RobertButtigieg** [comment]
> I don't want and need Azure dependencies in my artifacts.
> So I'm not happy when this gets merged. If this is relevant to some people, isn't it better to implement this as a separate module?

Hi @ArjanSchouten, that's why I have imported the 2 new azure libraries as optional dependencies so that the library won't include any new dependencies unless you specifically specify them yourself.

File: core/src/main/java/com/onelogin/saml2/settings/Saml2Settings.java [hunk 1]
```diff
@@ -911,12 +930,8 @@ private boolean checkIdpx509certRequired () {
 			LOGGER.error(errorMsg);
 		}
 
-		if ((this.getAuthnRequestsSigned() == true ||
-			  this.getLogoutRequestSigned() == true ||
-			  this.getLogoutResponseSigned() == true ||
-			  this.getWantAssertionsEncrypted() == true ||
-			  this.getWantNameIdEncrypted() == true)
-			  && this.checkSPCerts() == false) {
+		if (this.getHsm() == null && (this.getAuthnRequestsSigned() || this.getLogoutRequestSigned()
```
Comment (pitbulk) [inline core/src/main/java/com/onelogin/saml2/settings/Saml2Settings.java:933]: Replace by:
```
if ((this.getAuthnRequestsSigned() == true ||
			  this.getLogoutRequestSigned() == true ||
			  this.getLogoutResponseSigned() == true ||
			  this.getWantAssertionsEncrypted() == true ||
			  this.getWantNameIdEncrypted() == true)
			  && !this.checkSPCerts() && this.getHsm() == null) {
```

File: core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java [hunk 2]
```diff
@@ -1119,7 +1123,13 @@ private Document decryptAssertion(Document dom) throws XPathExpressionException,
 		    throw new ValidationError("No /samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData element found", ValidationError.MISSING_ENCRYPTED_ELEMENT);
 		}
 		Element encryptedData = (Element) encryptedDataNodes.item(0);
-		Util.decryptElement(encryptedData, key);
+
+		if (hsm != null) {
+			Util.decryptUsingHsm(encryptedData, hsm);
```
Comment (pitbulk) [inline core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:1128]: can you rename to decryptElementUsingHsm ?

**Comment by pitbulk** [comment]
I left 2 minor changes, and I believe we are ready to go.

@ArjanSchouten, I agree with @RobertButtigieg. The feature dependencies are marked as [optional](https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html) so if you don't require them, not gonna be part of your project.

**Comment by ArjanSchouten** [comment]
@pitbulk first of all, thanks for your quick response. 
For the optional dependency: I see now, thanks 😅 .

I think the HSM abstract class (IMHO it can be an interface) is a core thing. Different implementations such as AzureKeyVault can be in a different module. Since its an optional dependency, my original concern has been addressed. Creating a module for 1 class and 2 dependencies is maybe not worth it 😄 .

Thanks 👍

File: core/src/main/java/com/onelogin/saml2/settings/Saml2Settings.java (hunk 1 above)
Comment (RobertButtigieg) [reply core/src/main/java/com/onelogin/saml2/settings/Saml2Settings.java:933]: Hi @pitbulk 

First of all, sorry for my late reply! I have committed the requested changes on my forked repository. Shall you re-open this pull request or shall I open a new one in order to merge these changes?

**Pull request merged**